File: internal/adminapi/reconcile.go
package adminapi

import (
	"encoding/json"
	"errors"
	"io"
	"net/http"

	"github.com/danchupin/strata/internal/auth"
	"github.com/danchupin/strata/internal/meta"
	"github.com/danchupin/strata/internal/s3api"
)

// ReconcileJSON is the operator-console wire shape for a reconcile job (US-006).
// It mirrors the s3api admin endpoint (adminReconcileResponse) but rides the
// cookie-authenticated /admin/v1 surface the web console talks to, and renders
// the timestamps as Unix seconds so the UI can show an elapsed clock the same
// way DrainProgressBar / the reshard panel do.
//
// State is one of "queued" | "running" | "done" | "error" (meta.ReconcileState*).
// A job carries BOTH the orphan-pass counters (Scanned/Orphans*/AbsentBackref)
// and the dangling-pass counters (ManifestsScanned/Healthy/Dangling*); which set
// is meaningful is keyed off Bucket (empty == orphan pass, non-empty == dangling
// pass) — the console renders the relevant block.
type ReconcileJSON struct {
	OK            bool   `json:"ok"`
	ID            string `json:"id"`
	Cluster       string `json:"cluster,omitempty"`
	Pool          string `json:"pool,omitempty"`
	Namespace     string `json:"namespace,omitempty"`
	Bucket        string `json:"bucket,omitempty"`
	Policy        string `json:"policy"`
	State         string `json:"state"`
	Cursor        string `json:"cursor,omitempty"`
	Scanned       int64  `json:"scanned"`
	OrphansFound   int64 `json:"orphans_found"`
	OrphansGC      int64 `json:"orphans_gc"`
	OrphansReport  int64 `json:"orphans_report"`
	OrphansRestore int64 `json:"orphans_restore"`
	AbsentBackref  int64 `json:"absent_backref"`
	// Dangling-pass (US-003) counters.
	ManifestsScanned   int64  `json:"manifests_scanned"`
	Healthy            int64  `json:"healthy"`
	DanglingFound      int64  `json:"dangling_found"`
	DanglingQuarantine int64  `json:"dangling_quarantine"`
	DanglingReport     int64  `json:"dangling_report"`
	DanglingDelete     int64  `json:"dangling_delete"`
	Errors             int64  `json:"errors"`
	Message            string `json:"message,omitempty"`
	StartedAt          int64  `json:"started_at,omitempty"`
	UpdatedAt          int64  `json:"updated_at,omitempty"`
}

// reconcilePostRequest is the JSON body accepted by POST /admin/v1/reconcile.
// Two passes, discriminated by Bucket:
//   - ORPHAN (data->meta): Bucket empty. Cluster + Pool required; Namespace
//     optional. Policy is report (default), gc, or restore (US-002b — rebuild
//     the manifest row from the back-reference).
//   - DANGLING (meta->data): Bucket set (the bucket NAME, resolved to its UUID
//     here). Cluster/Pool ignored. Policy is report (default), quarantine, or
//     delete (US-003b — GC the version's chunks + remove the row).
type reconcilePostRequest struct {
	Cluster   string `json:"cluster"`
	Pool      string `json:"pool"`
	Namespace string `json:"namespace"`
	Bucket    string `json:"bucket"`
	Policy    string `json:"policy"`
}

func reconcileJSON(job *meta.ReconcileJob) ReconcileJSON {
	return ReconcileJSON{
		OK:                 true,
		ID:                 job.ID,
		Cluster:            job.Cluster,
		Pool:               job.Pool,
		Namespace:          job.Namespace,
		Bucket:             job.Bucket,
		Policy:             job.Policy,
		State:              job.State,
		Cursor:             job.Cursor,
		Scanned:            job.Scanned,
		OrphansFound:       job.OrphansFound,
		OrphansGC:          job.OrphansGC,
		OrphansReport:      job.OrphansReport,
		OrphansRestore:     job.OrphansRestore,
		AbsentBackref:      job.AbsentBackref,
		ManifestsScanned:   job.ManifestsScanned,
		Healthy:            job.Healthy,
		DanglingFound:      job.DanglingFound,
		DanglingQuarantine: job.DanglingQuarantine,
		DanglingReport:     job.DanglingReport,
		DanglingDelete:     job.DanglingDelete,
		Errors:             job.Errors,
		Message:            job.Message,
		StartedAt:          unixOrZero(job.CreatedAt),
		UpdatedAt:          unixOrZero(job.UpdatedAt),
	}
}

// handleReconcileStart serves POST /admin/v1/reconcile. It queues a reconcile
// pass and returns 202 immediately — the leader-elected `reconcile` worker
// drains the job out of band (a live-cluster scan must never block the request
// goroutine). The console polls GET /admin/v1/reconcile/{id} to watch it
// converge, mirroring the reshard-progress UX. Audit row admin:Reconcile.
func (s *Server) handleReconcileStart(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()
	if s.Meta == nil {
		writeJSONError(w, http.StatusInternalServerError, "Internal", "meta store not configured")
		return
	}
	body, err := io.ReadAll(io.LimitReader(r.Body, 4<<10))
	if err != nil {
		writeJSONError(w, http.StatusBadRequest, "BadRequest", "failed to read body")
		return
	}
	var req reconcilePostRequest
	if len(body) > 0 {
		if jerr := json.Unmarshal(body, &req); jerr != nil {
			writeJSONError(w, http.StatusBadRequest, "MalformedRequest", "invalid JSON: "+jerr.Error())
			return
		}
	}
	policy := req.Policy
	if policy == "" {
		policy = meta.ReconcilePolicyReport
	}

	var (
		bucketID      string
		auditResource string
	)
	if req.Bucket != "" {
		// Dangling pass: resolve the bucket name to its UUID for the worker.
		b, gerr := s.Meta.GetBucket(r.Context(), req.Bucket)
		if gerr != nil {
			if errors.Is(gerr, meta.ErrBucketNotFound) {
				writeJSONError(w, http.StatusNotFound, "NoSuchBucket", "bucket not found")
				return
			}
			writeJSONError(w, http.StatusInternalServerError, "Internal", gerr.Error())
			return
		}
		bucketID = b.ID.String()
		auditResource = "bucket:" + req.Bucket
	} else {
		if req.Cluster == "" || req.Pool == "" {
			writeJSONError(w, http.StatusBadRequest, "InvalidArgument",
				"cluster and pool are required (or pass bucket for a dangling-manifest pass)")
			return
		}
		auditResource = "cluster:" + req.Cluster + "/" + req.Pool
	}

	ctx := r.Context()
	owner := auth.FromContext(ctx).Owner
	s3api.SetAuditOverride(ctx, "admin:Reconcile", auditResource, "", owner)

	job, err := s.Meta.StartReconcile(ctx, req.Cluster, req.Pool, req.Namespace, bucketID, policy)
	if err != nil {
		if errors.Is(err, meta.ErrReconcileInvalidPolicy) {
			writeJSONError(w, http.StatusBadRequest, "InvalidArgument", err.Error())
			return
		}
		writeJSONError(w, http.StatusInternalServerError, "Internal", err.Error())
		return
	}
	writeJSON(w, http.StatusAccepted, reconcileJSON(job))
}

// handleReconcileStatus serves GET /admin/v1/reconcile/{id}. It reads a job by
// id so the console can poll a pass to completion and render the post-run
// orphan / dangling summary.
func (s *Server) handleReconcileStatus(w http.ResponseWriter, r *http.Request) {
	if s.Meta == nil {
		writeJSONError(w, http.StatusInternalServerError, "Internal", "meta store not configured")
		return
	}
	id := r.PathValue("id")
	if id == "" {
		writeJSONError(w, http.StatusBadRequest, "BadRequest", "job id is required")
		return
	}
	job, err := s.Meta.GetReconcileJob(r.Context(), id)
	if errors.Is(err, meta.ErrReconcileNotFound) {
		writeJSONError(w, http.StatusNotFound, "NoSuchReconcileJob", "no reconcile job with that id")
		return
	}
	if err != nil {
		writeJSONError(w, http.StatusInternalServerError, "Internal", err.Error())
		return
	}
	writeJSON(w, http.StatusOK, reconcileJSON(job))
}

File: web/src/api/reconcile.ts
// Reconcile-job wrappers for /admin/v1/reconcile (US-006). Wire shape mirrors
// adminapi.ReconcileJSON in internal/adminapi/reconcile.go.
//
// The console mirrors the drain / reshard-progress UX: POST queues a job (202),
// the leader-elected `reconcile` worker drains it out of band, and the GET
// endpoint is polled to watch the pass converge (queued → running → done).
//
// Two passes, discriminated by which params are set:
//   - ORPHAN  (data→meta): cluster + pool (+ optional namespace). Policy is
//     report (default), gc, or restore. Walks the RADOS pool, attributes each
//     chunk to an owner via its back-reference, flags chunks no manifest
//     references. restore (US-002b) rebuilds the manifest row from the
//     back-reference for a genuinely-absent version (plaintext-only).
//   - DANGLING (meta→data): bucket (a bucket NAME). Policy is report (default)
//     or quarantine. Walks every object version's manifest and probes each
//     chunk; a manifest with a missing chunk is dangling.
//
// rebuild-index is intentionally NOT here — it is a destructive last-resort
// CLI-only op (`strata admin rebuild-index`) gated behind shell access; the
// console links to the runbook instead of exposing a one-click rebuild.

export type ReconcileState = 'queued' | 'running' | 'done' | 'error';

export type ReconcilePass = 'orphan' | 'dangling';

// Orphan-pass policies. restore (US-002b) rebuilds the manifest row from the
// back-reference for a genuinely-absent version.
export type OrphanPolicy = 'report' | 'gc' | 'restore';
// Dangling-pass policies.
export type DanglingPolicy = 'report' | 'quarantine';

export interface ReconcileJob {
  ok: boolean;
  id: string;
  cluster?: string;
  pool?: string;
  namespace?: string;
  bucket?: string;
  policy: string;
  state: ReconcileState;
  cursor?: string;
  scanned: number;
  orphans_found: number;
  orphans_gc: number;
  orphans_report: number;
  orphans_restore: number;
  absent_backref: number;
  manifests_scanned: number;
  healthy: number;
  dangling_found: number;
  dangling_quarantine: number;
  dangling_report: number;
  errors: number;
  message?: string;
  started_at?: number;
  updated_at?: number;
}

export interface StartReconcileRequest {
  cluster?: string;
  pool?: string;
  namespace?: string;
  bucket?: string;
  policy?: string;
}

// extractErrorMessage pulls the server's JSON {code,message} so the form can
// render the inline error (e.g. 400 InvalidArgument on a missing pool, 404
// NoSuchBucket on a dangling pass against a missing bucket).
async function extractErrorMessage(resp: Response): Promise<string> {
  try {
    const j = (await resp.json()) as { message?: string; code?: string };
    if (j.message) return j.message;
    if (j.code) return j.code;
  } catch {
    // ignore JSON parse failure
  }
  return `${resp.status} ${resp.statusText}`;
}

// startReconcile queues a reconcile pass and returns the queued job descriptor
// (202). Throws on 4xx/5xx with the server message so the form renders the
// inline error.
export async function startReconcile(
  req: StartReconcileRequest,
): Promise<ReconcileJob> {
  const resp = await fetch('/admin/v1/reconcile', {
    method: 'POST',
    credentials: 'same-origin',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  });
  if (!resp.ok) {
    throw new Error(`start reconcile: ${await extractErrorMessage(resp)}`);
  }
  return (await resp.json()) as ReconcileJob;
}

// fetchReconcileJob reads a reconcile job by id so the console can poll a pass
// to completion and render the post-run orphan / dangling summary.
export async function fetchReconcileJob(id: string): Promise<ReconcileJob> {
  const resp = await fetch(`/admin/v1/reconcile/${encodeURIComponent(id)}`, {
    method: 'GET',
    credentials: 'same-origin',
  });
  if (!resp.ok) {
    throw new Error(`reconcile status: ${await extractErrorMessage(resp)}`);
  }
  return (await resp.json()) as ReconcileJob;
}
